### ISSUE TITLE
Jdiwbsb

### ISSUE THREAD

**Issue by davalillo-rm** [davalillo-rm/davalillo-rm#151]
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
import os

print("Iniciando el Cruce Maestro: Modelo Forward + Shocks PCA...")

# =====================================================================
# 1. CARGA DE LAS DOS FUENTES OFICIALES
# =====================================================================
# Cambia las rutas si es necesario
ruta_backtest = r'C:\Users\s7398771\OneDrive - The Bank of Nova Scotia\Documentos\CDBCRP\Backtesting_Real.xlsx'
ruta_wide = r'C:\Users\s7398771\OneDrive - The Bank of Nova Scotia\Documentos\CDBCRP\Base_WIDE_Fuente_PCA.xlsx'
ruta_salida = r'C:\Users\s7398771\OneDrive - The Bank of Nova Scotia\Documentos\CDBCRP\Modelo_Final_Arbitraje.xlsx'

df_backtest = pd.read_excel(ruta_backtest)
df_backtest['Fecha'] = pd.to_datetime(df_backtest['Fecha'])

df_wide = pd.read_excel(ruta_wide)
df_wide['FECHA'] = pd.to_datetime(df_wide['FECHA'])
df_wide = df_wide.set_index('FECHA')

# =====================================================================
# 2. CÁLCULO MASIVO DE SHOCKS SISTÉMICOS (EL MOTOR PCA)
# =====================================================================
# Filtramos solo las columnas de tasas base y forzamos el orden correcto
orden_nodos = ['CD1M_TASA_BASE', 'CD3M_TASA_BASE', 'CD6M_TASA_BASE', 'CD1Y_TASA_BASE']
df_pca_input = df_wide[orden_nodos]

# Calculamos diferencias diarias para la estacionariedad
df_diff = df_pca_input.diff().dropna()

# Entrenamos el PCA y reconstruimos toda la historia de una sola vez
pca = PCA(n_components=3)
scores_historicos = pca.fit_transform(df_diff)
shocks_reconstruidos = pca.inverse_transform(scores_historicos)

# Convertimos la matriz de shocks puros a un DataFrame fácil de leer
df_shocks = pd.DataFrame(
    shocks_reconstruidos, 
    index=df_diff.index, 
    columns=['Shock_1M', 'Shock_3M', 'Shock_6M', 'Shock_1Y']
)

# =====================================================================
# 3. EMPAREJAMIENTO DE BUCKETS (MATCHING LOGIC)
# =====================================================================
# Le decimos a cada bono histórico qué columna de shock le corresponde mirar
def asignar_bucket_shock(dias):
    if 21 <= dias <= 40: return 'Shock_1M'
    elif 80 <= dias <= 100: return 'Shock_3M'
    elif 170 <= dias <= 190: return 'Shock_6M'
    elif 340 <= dias <= 380: return 'Shock_1Y'
    return None

df_backtest['Columna_Shock'] = df_backtest['Dias'].apply(asignar_bucket_shock)

# Filtramos los que no entran en el análisis (ej. menores a 21 días)
df_backtest = df_backtest.dropna(subset=['Columna_Shock']).copy()

# =====================================================================
# 4. LA FUSIÓN (MERGE) Y CÁLCULO DEL ARBITRAJE
# =====================================================================
# Unimos el backtesting con los shocks macroeconómicos del mismo día
df_final = pd.merge(
    df_backtest, 
    df_shocks, 
    left_on='Fecha', 
    right_index=True, 
    how='inner'
)

# Extraemos dinámicamente el valor del shock según el bucket del bono
df_final['Shock_Mercado'] = df_final.apply(lambda fila: fila[fila['Columna_Shock']], axis=1)

# LA FÓRMULA FINAL DE INEFICIENCIA
# Arbitraje Puro = Diferencia del Forward - Shock Sistémico del Mercado
df_final['Arbitraje_Puro'] = df_final['Diferencia'] - df_final['Shock_Mercado']

# =====================================================================
# 5. LIMPIEZA Y EXPORTACIÓN
# =====================================================================
columnas_reporte = [
    'Fecha', 'ISIN', 'Dias', 'Yield', 'TASA_Manana', '1d_Forward', 
    'Diferencia', 'Columna_Shock', 'Shock_Mercado', 'Arbitraje_Puro'
]

df_reporte = df_final[columnas_reporte].sort_values(by=['Fecha', 'Dias']).copy()

# Guardamos el Excel definitivo
os.makedirs(os.path.dirname(ruta_salida), exist_ok=True)
df_reporte.to_excel(ruta_salida, index=False, sheet_name='Arbitraje_PCA')

print(f"\n¡Éxito total! Se han cruzado {len(df_reporte)} registros históricos.")
print(f"Tu insumo final ha sido guardado en:\n{ruta_salida}")

# Mostramos un mini-resumen de los hallazgos en consola
top_oportunidades = df_reporte.reindex(df_reporte['Arbitraje_Puro'].abs().sort_values(ascending=False).index).head(5)
print("\n--- TOP 5 MAYORES DESVIACIONES IDIOSINCRÁTICAS (BPS) ---")
print(top_oportunidades[['Fecha', 'ISIN', 'Diferencia', 'Shock_Mercado', 'Arbitraje_Puro']])
